File: 2024/10_go/partAB.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

var nextSteps [][]int = [][]int{
	{1, 0},
	{0, 1},
	{-1, 0},
	{0, -1}}

func inMap(r, c, lr, lc int) bool {
	return (r >= 0) && (c >= 0) && (r < lr) && (c < lc)
}

func removeDuplicates(input [][]int) [][]int {
	seen := make(map[int]bool)
	result := [][]int{}

	for _, elem := range input {
		if !seen[elem[0]*10000+elem[1]] {
			result = append(result, elem)
			seen[elem[0]*10000+elem[1]] = true
		}
	}

	return result
}

func runTrail(mapa [][]int, start []int) [][]int {
	lr := len(mapa)
	lc := len(mapa[0])
	startValue := mapa[start[0]][start[1]]
	if startValue == 9 {

		return [][]int{start}
	}

	validPath := [][]int{}
	for _, d := range nextSteps {
		newPos := []int{start[0] + d[0], start[1] + d[1]}
		if inMap(newPos[0], newPos[1], lr, lc) {
			if mapa[newPos[0]][newPos[1]] == startValue+1 {
				valids := runTrail(mapa, newPos)
				validPath = append(validPath, valids...)
			}
		}
	}
	return validPath
}

func main() {
	start := time.Now()
	var data, _ = os.ReadFile("myinput")
	var lines = strings.Split(string(data), "\n")

	var mapa [][]int
	for _, l := range lines {

		row := []int{}
		for _, c := range l {
			num, _ := strconv.Atoi(string(c))
			row = append(row, num)
		}

		mapa = append(mapa, row)
	}

	scores := 0
	scoresA := 0
	for r, l := range mapa {
		for c, val := range l {
			if val == 0 {
				options := runTrail(mapa, []int{r, c})
				scores += len(options)
				scoresA += len(removeDuplicates(options))
			}
		}
	}
	fmt.Printf("ResultadoA : %d\n", scoresA)
	fmt.Printf("ResultadoB : %d\n", scores)
	fmt.Printf("Time: %s\n", time.Since(start))

}

File: 2024/10/partAB.py
import datetime

startTime = datetime.datetime.now()

mapa = [list(l.replace("\n","")) for l in open("myinput").readlines()]
for r in range(len(mapa)):
    for c in range(len(mapa)):
        try:
            mapa[r][c] = int(mapa[r][c])
        except ValueError as e:
            mapa[r][c] = -1


def debugPrint(m):
    print("  ", end="")
    for i in range(len(m[0])):
        print(str(i) + " ", end="")
    print()
    cnt = 0
    for i in m:
        print(str(cnt)+" ", end = "")
        cnt += 1
        for j in i:
            print(str(j)+ " ", end="")
        print()

lc = len(mapa[0])
lr = len(mapa[0])
def inMap(r, c):
    return ( r >= 0) and (c>=0) and (c<lc) and (r < lr)

nextSteps =[
    ( 1, 0),
    ( 0, 1),
    (-1, 0),
    (0, -1)
]

def runTrail(mapa, start):
    startValue = mapa[start[0]][start[1]]

    if startValue == 9:
        return [start]

    validTrails = []
    for d in nextSteps:
        newStep = (start[0] + d[0], start[1] + d[1])
        if inMap(newStep[0], newStep[1]):
            if mapa[newStep[0]][newStep[1]] == startValue + 1:
                valids = runTrail(mapa, newStep)
                validTrails += valids

    return validTrails


scores = 0
scoresB = 0
for r in range(len(mapa)):
    for c in range(len(mapa)):
        if mapa[r][c] == 0:
            ret = runTrail(mapa, (r, c))
            scoresB += len(ret)
            scores += len(set(ret))

b = datetime.datetime.now()
print(f"Result: {scores}")
print(f"ResultB: {scoresB}")
print(f"Time: {(b-startTime).microseconds}")
